ipaserver/install/ipa_migrate.py

```diff
@@ -0,0 +1,2028 @@
+  it is added, otherwise we compare the remote and local entries and update the
+  local entry --> update_local_entry()
+    - Normalize attribute names to match the case of the local server's attrs
+    - Loop over remote entry attributes
+        - Skipping attrs from the "ignore list"
+        - Check the migration mode (prod-mode & stage-mode)
+            - If prod-mode, we migrate SIDs and DNA ranges
+            - If stage-mode, SIDs and DNA are skipped, and dna attributes
+              (uidNumber, gidNumber) are reset to the magic value
+        - Check if attribute is defined in the mappings "special_attrs" list
+            - If it is a special attribute then handle attribute comparison
+              according to the special definition (e.g. list) and update the
+              local entry. Then move on to next attribute...
+        - If the attribute is not "special" then we simply compare attribute to
+          attribute.
+            - If requested, DNA values are reset (magic regen) at this stage
+            - If attribute being updated is "single-valued" then "replace" the
+              value.  If its "multi-valued" then "append" the different value.
+
+Other
+--------------------------------
+There is a lot of normalization going on because we use dictionaries for
+attribute names, but attribute names are CIS, we have to normalize them during
+comparison, but we need to continue to use the original case when we go to
+update the local entry.  So in some cases we normalize to all lowercase, but
+when updatng the local entry we normalize the case of the remote attributes to
+match the local entry's attribute case
+
+
+What's next
+ - ask trivino to skip teardown on CI so I can get data
+
+ Some users/customers add their own entries to the db. Need more info on this
+ as those entries will not be migrated by default
+
+ after password change (but doesn't look like we can reset admin pass:
+     kinit_as_user -> for admin  in IPA API somewhere
+
+ write test from "integration" tests (new class)
+"""
+
+
+logger = logging.getLogger(__name__)
+
+VERSION = "1.0"
+
+# Audit stats
+stats = {
+    # Schema
+    'schema_attrs_added': 0,
+    'schema_attrs_skipped': 0,
+    'schema_oc_added': 0,
+    'schema_oc_skipped': 0,
+    'schema_processed': 0,
+    # Config
+    'config_processed': 0,
+    'config_migrated': 0,
+    # general
+    'conflicts': 0,  # LDIF log entries
+    'ignored_errors': 0,  # force option
+    # db
+    'reset_range': 0,
+    'custom': 0,
+    'total_db_entries': 0,
+    'total_db_migrated': 0,
+}
+
+
+#
+# Generic helper functions
+#
+def normalize_attr(entry_attrs, attr):
+    """
+    Convert all the entry attributes that match "attr" to same case as "attr"
+    """
+    vals = []
+    nattr = attr.lower()
+    for key_attr in entry_attrs:
+        if key_attr.lower() == nattr:
+            vals = entry_attrs[key_attr].copy()
+            del entry_attrs[key_attr]
+            break
+    if len(vals) > 0:
```

Comment:
I think `if vals` is probably more pythonic. An empty list will be treated as false in this case.